File: ray/ray.go
package ray

import "github.com/sagnikc395/rayt/vec3"

type Point3 = vec3.Vec3
type Ray struct {
	Orig Point3
	Dir  vec3.Vec3
}

func NewRay(origin Point3, direction vec3.Vec3) *Ray {
	return &Ray{
		Orig: origin,
		Dir:  direction,
	}
}

func (r *Ray) Origin() Point3 {
	return r.Orig
}

func (r *Ray) Direction() *vec3.Vec3 {
	return &r.Dir
}

func (r *Ray) At(t float64) Point3 {
	return Point3{E: vec3.VecAdd(&r.Orig, r.Dir.ScaleUp(t)).E}
}

func (r *Ray) RayColor() vec3.Vec3 {
	if r.HitSphere(vec3.NewVec3(0, 0, -1), 0.5) {
		return *vec3.NewVec3(1, 0, 0)
	}
	unit_direction := r.Dir.UnitVector()
	a := 0.5 * (unit_direction.Y() + 1.0)

	exp1 := vec3.NewVec3(1.0, 1.0, 1.0)
	exp2 := vec3.NewVec3(0.5, 0.7, 1.0)

	return *vec3.VecAdd(exp1.ScaleUp(1.0-a), exp2.ScaleUp(a))
}

func (r *Ray) HitSphere(center *vec3.Vec3, radius float64) bool {
	oc := vec3.VecSub(center, &r.Orig)
	a := vec3.DotProduct(&r.Dir, &r.Dir)
	b := -2.0 * vec3.DotProduct(&r.Dir, oc)
	c := vec3.DotProduct(oc, oc) - (radius * radius)
	discriminant := b*b - (4 * a * c)
	return discriminant >= 0
}

File: src/ray.rs
use crate::vec3::{dot, Vec3};

pub type Point3 = Vec3;

#[derive(Debug, Clone, Copy)]
pub struct Ray {
    orig: Point3,
    dir: Vec3,
}

impl Ray {
    //constructor with parameters
    pub fn new(origin: Point3, direction: Vec3) -> Self {
        Self {
            orig: origin,
            dir: direction,
        }
    }
    //getter for origin and direction
    pub fn origin(&self) -> Point3 {
        self.orig
    }
    pub fn direction(&self) -> Vec3 {
        self.dir
    }
    //calc point at param t
    pub fn at(self, t: f64) -> Point3 {
        self.orig + self.dir * t
    }
}

pub fn check_hit_sphere(center: Point3, radius: f64, r: &Ray) -> bool {
    // check if it hits the sphere or not
    let oc = center - r.origin();
    let a = dot(r.direction(), r.direction());
    let b = -2.0 * dot(r.direction(), oc);
    let c = dot(oc, oc) - radius * radius;
    let discriminant = b * b - (4.0 * a * c);
    discriminant >= 0.0
}

//although in the text book , they have modified the existing implementation hit_sphere to return closest hit point
// here i have abstracted into a different function
pub fn hit_sphere(center: Point3, radius: f64, r: &Ray) -> f64 {
    let oc = center - r.origin();
    let a = dot(r.direction(), r.direction());
    let b = -2.0 * dot(r.direction(), oc);
    let c = dot(oc, oc) - radius * radius;
    let discriminant = b * b - (4.0 * a * c);

    if discriminant < 0.0 {
        return -1.0;
    } else {
        return (-b - discriminant.sqrt()) / 2.0 * a;
    }
}
